File: python/src/tdcv2/output/parquet_output.py
"""A run written as a typed binary file instead of as text.

The same preparation as the text renderer — one engine, one registry — so a config exported to
Parquet holds exactly the data it would have printed for that seed. Only the serialisation differs:
instead of formatting a record, each named ``<data>`` becomes a typed column.

Rows go out in row groups, each built, written and released before the next one starts, so memory
stays bounded however large the run is.
"""

from __future__ import annotations

from collections.abc import Callable, Iterator
from dataclasses import dataclass

from ..format import interpolate
from ..model.config import Config
from . import column_type
from . import columns as columns_lib
from .column_type import ColumnType
from .parquet import convert, writer

# Rows per row group. Bounds peak memory and lets a reader skip whole groups. It is also the unit
# parallel generation would split on, because a group's bytes do not depend on where it sits.
ROW_GROUP_ROWS = 50_000

# An untyped column is text. Never guess from the values — a string never corrupts data.
DEFAULT_TYPE = column_type.parse("string")


@dataclass(frozen=True, slots=True)
class _Plan:
    """Everything decided before a single row is rendered."""

    declared: list[columns_lib.Declared]
    columns: list[writer.Column]
    types: list[ColumnType]
    separators: list[str | None]


def write(config: Config, rows, out: Callable[[bytes], None], on_progress=None) -> None:
    """The run written to a sink as a ``.parquet`` file."""
    plan = _plan(config)
    writer.write(plan.columns, _batches(plan, config, rows, on_progress), out)


def to_bytes(config: Config, rows) -> bytes:
    """The same, in memory — for a small output and for tests."""
    out = bytearray()
    write(config, rows, out.extend)
    return bytes(out)


def schema_of(config: Config) -> list[writer.Column]:
    """The resolved schema, for telling the user which types were chosen."""
    return _plan(config).columns


def _plan(config: Config) -> _Plan:
    declared: list[columns_lib.Declared] = []
    for line in config.block:
        for part in line.parts:
            name = None if part.name is None else part.name.strip()
            if not name:
                continue  # decorative text, not a column
            type_ = None
            if part.type is not None:
                try:
                    type_ = column_type.parse_output(part.type)
                except ValueError as e:
                    raise ValueError(f'column "{name}": {e}') from None
            declared.append(columns_lib.Declared(name, part.text, type_))

    if not declared:
        raise ValueError(
            'Parquet output needs at least one named column — add name="…" to a <data> in the '
            "<block>"
        )
    columns_lib.check_unique(declared)

    types: list[ColumnType] = []
    separators: list[str | None] = []
    built: list[writer.Column] = []
    for column in declared:
        type_ = columns_lib.resolve(column, config) or DEFAULT_TYPE
        types.append(type_)
        # A declared []T needs a separator too; a comma when the column was typed by hand rather
        # than derived from a repeating generator.
        if type_.is_list:
            source = columns_lib.sole_reference(column.template, config.inject)
            separator = None if source is None else columns_lib.separator_of(source, config)
            separators.append(separator or ",")
        else:
            separators.append(None)
        built.append(writer.Column(column.name, type_))
    return _Plan(declared, built, types, separators)


def _batches(
    plan: _Plan, config: Config, rows, on_progress=None
) -> Iterator[list[list[writer.Cell]]]:
    count = rows.count if isinstance(rows.count, int) else rows.count()
    start = 0
    while start < count:
        # Once per row group, which is fifty thousand rows: coarser than the text path's
        # half-percent, and it has to be — a row group is the unit this writer works in, and
        # there is no moment inside one where a partial group means anything.
        if on_progress is not None:
            on_progress("render", start, count)
        end = min(start + ROW_GROUP_ROWS, count)
        batch: list[list[writer.Cell]] = [[] for _ in plan.columns]

        for row in range(start, end):
            lookup = _lookup(rows, row)
            for i, column in enumerate(plan.declared):
                text = interpolate.apply(column.template, config.inject, lookup)
                type_ = plan.types[i]
                try:
                    if type_.is_list:
                        # An empty cell is an EMPTY LIST, not a list holding one blank — splitting
                        # "" on a comma would otherwise conjure a phantom element.
                        elements = [] if text == "" else text.split(plan.separators[i])
                        batch[i].append(writer.Cell(texts=elements))
                    else:
                        batch[i].append(writer.Cell(value=convert.value(text, type_)))
                except ValueError as e:
                    raise ValueError(f'column "{column.name}", row {row + 1}: {e}') from None
        start = end
        yield batch


def _lookup(rows, row: int):
    names = set(rows.sequence_names())

    def lookup(name: str) -> str | None:
        value = rows.value(name, row)
        if value is None and name not in names:
            return None
        return "" if value is None else value

    return lookup

File: csharp/Tdcv2/Output/ParquetOutput.cs
using Tdcv2.Engine;
using Tdcv2.Format;
using Tdcv2.Model;
// `Convert` is a Parquet class here and a BCL one everywhere else; the alias keeps the
// two apart at the one call site rather than renaming a type to dodge a collision.
using Tdcv2.Output.Parquet;
using ParquetConvert = Tdcv2.Output.Parquet.Convert;

namespace Tdcv2.Output;

/// <summary>
/// A run written as a typed binary file instead of as text.
/// </summary>
/// <remarks>
/// <para>
/// The same preparation as the text renderer — one engine, one registry — so a config exported to
/// Parquet holds exactly the data it would have printed for that seed. Only the serialisation
/// differs: instead of formatting a record, each named <c>&lt;data&gt;</c> becomes a typed column.
/// </para>
/// <para>
/// Rows go out in row groups, each built, written and released before the next one starts, so memory
/// stays bounded however large the run is.
/// </para>
/// </remarks>
public static class ParquetOutput
{
    /// <summary>
    /// Rows per row group.
    /// </summary>
    /// <remarks>
    /// Bounds peak memory and lets a reader skip whole groups. It is also the unit parallel
    /// generation would split on, because a group's bytes do not depend on where it sits in the file.
    /// </remarks>
    public const int RowGroupRows = 50_000;

    /// <summary>An untyped column is text. Never guess from the values — a string never corrupts data.</summary>
    private static readonly ColumnType DefaultType = ColumnType.Parse("string");

    /// <summary>Write the run to a sink as a <c>.parquet</c> file.</summary>
    public static void Write(
        Config config, IRowSource rows, Stream output, Engine.Progress? onProgress = null)
    {
        Plan plan = BuildPlan(config);
        Writer.Write(plan.Columns, Batches(plan, config, rows, onProgress), output);
    }

    /// <summary>The same, in memory — for a small output and for tests.</summary>
    public static byte[] ToBytes(Config config, IRowSource rows)
    {
        var output = new MemoryStream();
        Write(config, rows, output);
        return output.ToArray();
    }

    /// <summary>The resolved schema, for telling the user which types were chosen.</summary>
    public static IReadOnlyList<Writer.Column> SchemaOf(Config config) => BuildPlan(config).Columns;

    /// <summary>Everything decided before a single row is rendered.</summary>
    private sealed record Plan(
        IReadOnlyList<Columns.Declared> Declared,
        IReadOnlyList<Writer.Column> Columns,
        IReadOnlyList<ColumnType> Types,
        IReadOnlyList<string?> Separators);

    private static Plan BuildPlan(Config config)
    {
        var declared = new List<Columns.Declared>();
        foreach (Line line in config.Block)
        {
            foreach (DataPart part in line.Parts)
            {
                string? name = part.Name?.Trim();
                if (string.IsNullOrEmpty(name))
                {
                    continue; // decorative text, not a column
                }

                ColumnType? type = null;
                if (part.Type is not null)
                {
                    try
                    {
                        type = ColumnType.ParseOutput(part.Type);
                    }
                    catch (ArgumentException e)
                    {
                        throw new ArgumentException($"column \"{name}\": {e.Message}", e);
                    }
                }

                declared.Add(new Columns.Declared(name, part.Text, type));
            }
        }

        if (declared.Count == 0)
        {
            throw new ArgumentException(
                "Parquet output needs at least one named column — add name=\"…\" to a <data> in "
                + "the <block>");
        }

        Columns.CheckUnique(declared);

        var types = new List<ColumnType>(declared.Count);
        var separators = new List<string?>(declared.Count);
        var columns = new List<Writer.Column>(declared.Count);
        foreach (Columns.Declared column in declared)
        {
            ColumnType type = Columns.Resolve(column, config) ?? DefaultType;
            types.Add(type);
            // A declared []T needs a separator too; a comma when the column was typed by hand rather
            // than derived from a repeating generator.
            if (type.IsList)
            {
                string? source = Columns.SoleReference(column.Template, config.Inject ?? "${{%}}");
                string? separator = source is null ? null : Columns.SeparatorOf(source, config);
                separators.Add(separator ?? ",");
            }
            else
            {
                separators.Add(null);
            }

            columns.Add(new Writer.Column(column.Name, type));
        }

        return new Plan(declared, columns, types, separators);
    }

    private static Writer.Batches Batches(
        Plan plan, Config config, IRowSource rows, Engine.Progress? onProgress = null)
    {
        int count = rows.Count;
        int start = 0;

        return () =>
        {
            if (start >= count)
            {
                return null;
            }

            // Once per row group, which is fifty thousand rows: coarser than the text path's
            // half-percent, and it has to be — a row group is the unit this writer works in, and
            // there is no moment inside one where a partial group means anything.
            onProgress?.Invoke("render", start, count);
            int end = Math.Min(start + RowGroupRows, count);
            var batch = new List<List<Writer.Cell>>(plan.Columns.Count);
            for (int i = 0; i < plan.Columns.Count; i++)
            {
                batch.Add(new List<Writer.Cell>(end - start));
            }

            for (int row = start; row < end; row++)
            {
                for (int i = 0; i < plan.Declared.Count; i++)
                {
                    Columns.Declared column = plan.Declared[i];
                    string text = Interpolate.Apply(
                        column.Template, config.Inject, new RowLookup(rows, row));
                    ColumnType type = plan.Types[i];
                    try
                    {
                        if (type.IsList)
                        {
                            // An empty cell is an EMPTY LIST, not a list holding one blank —
                            // splitting "" on a comma would otherwise conjure a phantom element.
                            batch[i].Add(new Writer.Cell.Elements(
                                text.Length == 0
                                    ? Array.Empty<string>()
                                    : Split(text, plan.Separators[i]!)));
                        }
                        else
                        {
                            batch[i].Add(new Writer.Cell.Scalar(ParquetConvert.Of(text, type)));
                        }
                    }
                    catch (ArgumentException e)
                    {
                        throw new ArgumentException(
                            $"column \"{column.Name}\", row {row + 1}: {e.Message}", e);
                    }
                }
            }

            start = end;
            return batch;
        };
    }

    private sealed class RowLookup : Interpolate.ILookup
    {
        private readonly IRowSource _rows;
        private readonly int _row;

        internal RowLookup(IRowSource rows, int row)
        {
            _rows = rows;
            _row = row;
        }

        public bool Has(string name) =>
            _rows.Value(name, _row) is not null || _rows.SequenceNames.Contains(name);

        public string Value(string name) => _rows.Value(name, _row) ?? "";
    }

    /// <summary>A literal split, not a regular expression — the separator is a piece of data.</summary>
    private static List<string> Split(string text, string separator)
    {
        var result = new List<string>();
        int at = 0;
        while (true)
        {
            int next = text.IndexOf(separator, at, StringComparison.Ordinal);
            if (next < 0)
            {
                result.Add(text[at..]);
                return result;
            }

            result.Add(text[at..next]);
            at = next + separator.Length;
        }
    }
}
